python/cudf/cudf/core/cut.py

```diff
@@ -0,0 +1,197 @@
+    adj = (mx - mn) * 0.001
+    if right:
+        bins[0] -= adj
+    else:
+        bins[-1] += adj
+
+    if right and include_lowest:
+        bins[0] = bins[0] - 10 ** (-precision)
+
+    # adjust bin edges precision
+    bins = cupy.around(bins, precision)
+
+    # checking for the correct inclusivity values
+    if right:
+        closed = "right"
+    elif not right:
+        closed = "left"
+        left_inclusive = True
+
+    if labels is None:
+        # get labels for categories
+        interval_labels = IntervalIndex.from_breaks(bins, closed=closed)
+    elif labels is not False:
+        if not (is_list_like(labels)):
+            raise ValueError(
+                "Bin labels must either be False, None or passed in as a "
+                "list-like argument"
+            )
+        if ordered and len(set(labels)) != len(labels):
+            raise ValueError(
+                "labels must be unique if ordered=True; pass ordered=False for"
+                "duplicate labels"
+            )
+        else:
+            if len(labels) != len(bins) - 1:
+                raise ValueError(
+                    "Bin labels must be one fewer than the number of bin edges"
+                )
+            if not ordered and len(set(labels)) != len(labels):
+                interval_labels = cudf.CategoricalIndex(
+                    labels, categories=None, ordered=False
+                )
+            else:
+                interval_labels = (
+                    labels if len(set(labels)) == len(labels) else None
+                )
+
+    # get the left and right edges of the bins as columns
+    left_edges = as_column(bins[:-1:])
+    right_edges = as_column(bins[+1::])
+    # the input arr must be changed to the same type as the edges
+    input_arr = input_arr.astype(left_edges._dtype)
+    # get the indexes for the appropriate number
+    index_labels = label_bins(
+        input_arr, left_edges, left_inclusive, right_edges, right_inclusive
+    )
+    if index_labels.base_mask:
+        index_labels._base_mask = None
+
+    if labels is False:
+        # if labels is false we return the bin indexes
+        indx_arr = index_labels.values
+        return indx_arr
+
+    if labels is not None:
+        if labels is not ordered and len(set(labels)) != len(labels):
+            # when we have duplicate labels and ordered is False, we
+            # should allow duplicate categories
+            new_data = [interval_labels[i][0] for i in index_labels.values]
+            return cudf.CategoricalIndex(
+                new_data, categories=sorted(set(labels)), ordered=False
+            )
+    col = build_categorical_column(
+        categories=interval_labels, codes=index_labels, ordered=ordered
+    )
+    categorical_index = cudf.core.index.as_index(col)
+    if retbins:
+        # if retbins is true we return the bins as well
+        return categorical_index, bins
+    else:
+        return categorical_index
```

Comment:
Also just a note on this as well. Pandas returns a [Categorical](https://pandas.pydata.org/docs/reference/api/pandas.Categorical.html), but we don't have a `Categorical` method in cudf. @kkraus14 let me know if you'd like me to create one, or if its fine to stick with returning a `CategoricalIndex` 😄